parking_lot.py

```diff
@@ -0,0 +1,201 @@
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Car {exception.car_id} is invalid. Car plate cannot be null!!"})
+
+
+@app.exception_handler(CarAlreadyRemovedException)
+async def car_id_already_removed(exception: CarAlreadyRemovedException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Car {exception.car_id} already removed!"})
+
+@app.exception_handler(CarIdPatternException)
+async def car_id_invalid_pattern(exception: CarIdPatternException):
+    return JSONResponse(status_code=422, content={"status": "error", "message": f"Car {exception.car_id} is invalid!"})
+
+@app.get("/cars/")
+async def get_cars():
+    return PARKED_CARS
+
+
+@app.get("/cars/{car_id}", status_code=status.HTTP_200_OK)
+async def get_car(car_id: Optional[str] = None):
+    return PARKED_CARS[car_id]
+
+
+@app.post("/cars/", status_code=status.HTTP_201_CREATED)
+async def add_car(parked_car: ParkedCar):
+    if parked_car.car_id is None:
+        raise CarIdNullException(car_id=parked_car.car_id)
+
+    if parked_car.location <= 0 or parked_car.location > PARKING_LOT_CAPACITY:
+        raise InvalidLocationException(location=parked_car.location)
+
+    if parked_car.location in [x.get('location') for x in PARKED_CARS.values()]:
+        raise LocationNotAvailableException(location=parked_car.location)
+
+    start_date = parked_car.start.replace(tzinfo=None)
+    present = datetime.now()
+
+    """
+        In case the difference between the chosen date and the actual time is more
+        than the accepted time delay - by default 15 minutes - an error would be raised.
+        This is to prevent fraud or other mistakes that would decrease the fee unfairly.
+    """
+    if (start_date - present).total_seconds() > ACCEPTED_TIME_DELAY:
+        raise StartDateException(start=parked_car.start)
+
+    if parked_car.car_id in PARKED_CARS.keys():
+        raise CarIdExistsException(car_id=parked_car.car_id)
+
+    car_plate_pattern = re.compile(r"[A-Za-z0-9]+")
+    if not re.fullmatch(car_plate_pattern, parked_car.car_id):
+        raise CarIdPatternException(car_id=parked_car.car_id)
+
+    PARKED_CARS[parked_car.car_id] = {'status': 'parked', 'tariff': parked_car.tariff, 'location': parked_car.location, 'start': parked_car.start, 'end': None, 'fee': None}
+    return PARKED_CARS[parked_car.car_id]
+
+
+@app.put("/cars/{car_id}", status_code=status.HTTP_200_OK)
+async def remove_car(car_id: str, parked_car: ParkedCar):
+    if car_id not in PARKED_CARS.keys():
+        raise CarIdDoesNotExistException(car_id=car_id)
+
+    if PARKED_CARS[car_id].get('status') == 'removed':
+        raise CarAlreadyRemovedException(car_id=car_id)
+
+    tariff = PARKED_CARS[car_id].get('tariff')
+    start = dateutil.parser.isoparse(PARKED_CARS[car_id].get('start'))
+
+    PARKED_CARS[car_id] = {'status': 'removed', 'tariff': tariff, 'location': '', 'start': start, 'end': datetime.now(),
```

Comment:
MAJOR ISSUE: Service only increases PARKED_CARS (could be DB in general case) list with latest parking information of cars have ever parked. Over time millions of records will be there and impact overall service performance since PARKED_CARS is actively used everywhere.